src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -148,16 +148,25 @@ internal SocketError ReplaceHandle()
             // And put back the copied settings.  For DualMode, we use the value stored in the _handle
             // rather than querying the socket itself, as on Unix stacks binding a dual-mode socket to
             // an IPv6 address may cause the IPv6Only setting to revert to true.
-            if (_handle.IsTrackedOption(TrackedSocketOptions.DualMode)) DualMode = _handle.DualMode;
-            if (_handle.IsTrackedOption(TrackedSocketOptions.DontFragment)) DontFragment = dontFragment;
-            if (_handle.IsTrackedOption(TrackedSocketOptions.EnableBroadcast)) EnableBroadcast = broadcast;
-            if (_handle.IsTrackedOption(TrackedSocketOptions.LingerState)) LingerState = linger!;
-            if (_handle.IsTrackedOption(TrackedSocketOptions.NoDelay)) NoDelay = noDelay;
-            if (_handle.IsTrackedOption(TrackedSocketOptions.ReceiveBufferSize)) ReceiveBufferSize = receiveSize;
-            if (_handle.IsTrackedOption(TrackedSocketOptions.ReceiveTimeout)) ReceiveTimeout = receiveTimeout;
-            if (_handle.IsTrackedOption(TrackedSocketOptions.SendBufferSize)) SendBufferSize = sendSize;
-            if (_handle.IsTrackedOption(TrackedSocketOptions.SendTimeout)) SendTimeout = sendTimeout;
-            if (_handle.IsTrackedOption(TrackedSocketOptions.Ttl)) Ttl = ttl;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.DualMode)) DualMode = _handle.DualMode;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.DontFragment)) DontFragment = dontFragment;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.EnableBroadcast)) EnableBroadcast = broadcast;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.LingerState)) LingerState = linger!;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.NoDelay)) NoDelay = noDelay;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.ReceiveBufferSize)) ReceiveBufferSize = receiveSize;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.ReceiveTimeout)) ReceiveTimeout = receiveTimeout;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.SendBufferSize)) SendBufferSize = sendSize;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.SendTimeout)) SendTimeout = sendTimeout;
+            if (newHandle.IsTrackedOption(TrackedSocketOptions.Ttl)) Ttl = ttl;
+
+            SafeSocketHandle oldHandle = _handle;
+            Volatile.Write(ref _handle, newHandle);
+            oldHandle.Dispose();
+
+            if (Volatile.Read(ref _disposed) != 0)
+            {
+                _handle.Dispose();
```

Comment:
I think this can still race with `Socket.Dispose(bool)`, resulting in an abortive close

https://github.com/dotnet/runtime/blob/acb456435af158453aff1e8b485d6ba03afc975c/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.cs#L86-L89

instead of the socket's configuration-specific close, which is non-abortive by default:

https://github.com/dotnet/runtime/blob/acb456435af158453aff1e8b485d6ba03afc975c/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3197-L3202

@tmds do you think this is something we should worry about?

Reply:
We're closing the handle we've just created. A new connect attempt hasn't been started on this new handle yet.
I think we're good.